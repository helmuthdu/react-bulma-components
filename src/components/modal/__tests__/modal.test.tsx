import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { Modal, ModalCard, ModalCardBody, ModalCardFooter, ModalCardHeader, ModalCardTitle, ModalContent } from '..';

describe('Modal component', () => {
  it('should render modal-card-head', () => {
    const onClose = jest.fn();
    const { container, unmount } = render(
      <ModalCard onClose={onClose}>
        <ModalCardHeader onClose={onClose}>
          <ModalCardTitle>Modal Title</ModalCardTitle>
        </ModalCardHeader>
      </ModalCard>
    );
    fireEvent.click((container as any).querySelector('.delete'));
    expect(onClose).toHaveBeenCalledTimes(1);
    unmount();
  });

  it('should open the modal', () => {
    const onClose = jest.fn();
    const getComponent = (show: boolean) => (
      <Modal show={show} onClose={onClose}>
        <ModalCard>
          <ModalCardHeader>
            <ModalCardTitle>Modal Title</ModalCardTitle>
          </ModalCardHeader>
          <ModalCardBody>Model Content</ModalCardBody>
          <ModalCardFooter>Model Footer</ModalCardFooter>
        </ModalCard>
      </Modal>
    );
    const { rerender, unmount } = render(getComponent(false));
    expect(window.document.querySelector('div.modal is-active')).toBeNull();
    rerender(getComponent(true));
    expect(window.document.querySelector('div.modal is-active')).toMatchSnapshot();
    unmount();
  });

  it('should close the modal', () => {
    const onClose = jest.fn();
    const { unmount } = render(
      <Modal show onClose={onClose}>
        <ModalCard>
          <ModalCardHeader onClose={onClose}>
            <ModalCardTitle>Modal Title</ModalCardTitle>
          </ModalCardHeader>
        </ModalCard>
      </Modal>
    );

    fireEvent.click(window.document.querySelector('.modal-close') as HTMLButtonElement);
    expect(onClose).toHaveBeenCalledTimes(1);
    unmount();
  });

  it('should call close the modal on ESC key press', () => {
    const onClose = jest.fn();
    const { unmount } = render(
      <Modal show onClose={onClose}>
        <ModalCard>
          <ModalCardHeader onClose={onClose}>
            <ModalCardTitle>Modal Title</ModalCardTitle>
          </ModalCardHeader>
        </ModalCard>
      </Modal>
    );

    const event = new KeyboardEvent('keydown', { keyCode: 27 } as any);
    document.dispatchEvent(event);
    expect(onClose).toHaveBeenCalledTimes(1);
    unmount();
  });

  it('should not close the modal on ESC key press', () => {
    const onClose = jest.fn();
    const { unmount } = render(
      <Modal closeOnEsc={false} show onClose={onClose}>
        <ModalCard>
          <ModalCardHeader>
            <ModalCardTitle>Modal Title</ModalCardTitle>
          </ModalCardHeader>
        </ModalCard>
      </Modal>
    );

    const event = new KeyboardEvent('keydown', { keyCode: 27 } as any);
    document.dispatchEvent(event);
    expect(onClose).not.toHaveBeenCalled();
    unmount();
  });

  it('should not close the modal on other than ESC key press', () => {
    const onClose = jest.fn();
    const { unmount } = render(
      <Modal show onClose={onClose}>
        <ModalCard>
          <ModalCardHeader>
            <ModalCardTitle>Modal Title</ModalCardTitle>
          </ModalCardHeader>
        </ModalCard>
      </Modal>
    );

    const event = new KeyboardEvent('keydown', { keyCode: 23 } as any);
    document.dispatchEvent(event);
    expect(onClose).not.toHaveBeenCalled();
    unmount();
  });

  it('should render any child type', () => {
    const onClose = jest.fn();
    const { asFragment, unmount } = render(
      <Modal show onClose={onClose}>
        <div>CHILDREN</div>
        <div>CHILDREN</div>
      </Modal>
    );

    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  it('should close the modal if clicked on background', () => {
    const onClose = jest.fn();
    const { unmount } = render(
      <Modal show onClose={onClose} closeOnBlur>
        <ModalContent>Content</ModalContent>
      </Modal>
    );

    const background = window.document.querySelector('.modal-background') as HTMLDivElement;
    fireEvent.click(background);
    expect(onClose).toHaveBeenCalledTimes(1);
    unmount();
  });

  it('should not close the modal if clicked on background', () => {
    const onClose = jest.fn();
    const { unmount } = render(
      <Modal show onClose={onClose} closeOnBlur={false}>
        <ModalContent>Content</ModalContent>
      </Modal>
    );

    const background = window.document.querySelector('.modal-background') as HTMLDivElement;
    fireEvent.click(background);
    expect(onClose).not.toHaveBeenCalled();
    unmount();
  });

  it('should not show close button', () => {
    const onClose = jest.fn();
    const { asFragment, unmount } = render(
      <Modal showClose={false} show onClose={onClose}>
        <ModalContent>Content</ModalContent>
      </Modal>
    );
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  it('should render empty because no document on scope', () => {
    const onClose = jest.fn();
    const { asFragment, unmount } = render(
      <Modal show onClose={onClose}>
        <ModalContent>Content</ModalContent>
      </Modal>
    );
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });
});
