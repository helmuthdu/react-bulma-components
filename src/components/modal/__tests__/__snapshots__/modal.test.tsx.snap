// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should not show close button 1`] = `<DocumentFragment />`;

exports[`Modal component should open the modal 1`] = `null`;

exports[`Modal component should render any child type 1`] = `<DocumentFragment />`;

exports[`Modal component should render empty because no document on scope 1`] = `<DocumentFragment />`;
